src/NetAnalyzers/Core/Microsoft.NetCore.Analyzers/Runtime/ForwardCancellationTokenToAsyncMethods.cs

```diff
@@ -0,0 +1,244 @@
+                {
+                    if (methodBaseBody.SemanticModel.GetDeclaredSymbol(methodBaseBody.Syntax) is IMethodSymbol method)
+                    {
+                        declaration = method;
+                        // There are many kinds of blocks, so only break if we found a method symbol for this block.
+                        // Otherwise, blocks inside anonymous or local functions would not be detected - those would be the parent of the current operation.
+                        break;
+                    }
+                }
+
+                currentOperation = currentOperation.Parent;
+            }
+
+            return declaration != null;
+        }
+
+        // Check if the method only takes one ct and is the last parameter in the method signature.
+        // We want to compare the current method signature to any others with the exact same arguments in the exact same order.
+        private static bool VerifyMethodOnlyHasOneCancellationTokenAsLastArgument(
+            INamedTypeSymbol cancellationTokenType,
+            IMethodSymbol methodDeclaration,
+            out string? cancellationTokenParameterName)
+        {
+            cancellationTokenParameterName = null;
+
+            if (methodDeclaration.Parameters.Count(x => x.Type.Equals(cancellationTokenType)) == 1 &&
+                methodDeclaration.Parameters.Last() is IParameterSymbol lastParameter &&
+                lastParameter.Type.Equals(cancellationTokenType))
+            {
+                cancellationTokenParameterName = lastParameter.Name;
+            }
+
+            return cancellationTokenParameterName != null;
+        }
+
+        // Check if the currently used overload is the one that takes the ct, but is utilizing the default value offered in the method signature.
+        // We want to offer a diagnostic for this case, so the user explicitly passes the ancestor's ct.
+        private static bool InvocationIgnoresOptionalCancellationToken(INamedTypeSymbol cancellationTokenType, IInvocationOperation invocation, IMethodSymbol method)
+        {
+            if (method.Parameters.Any() && method.Parameters.Last() is IParameterSymbol lastParameter && lastParameter.Type.Equals(cancellationTokenType))
+            {
+                // If the ct parameter has a default value, return true if a value is not being explicitly passed in the invocation
+                return lastParameter.IsOptional && // Has a default value
+                       invocation.Arguments.Last() is IArgumentOperation lastArgument &&
+                       lastArgument.IsImplicit; // The default value is being used
+            }
+
+            return false;
+        }
+
+        // Check if there's a method overload with the same parameters as this one, in the same order, plus a ct at the end.
+        private static bool MethodHasCancellationTokenOverload(ITypeSymbol cancellationTokenType, IMethodSymbol method)
```

Comment:
@carlossanlop Seems like you can reuse the following existing extension method here: 

https://github.com/dotnet/roslyn-analyzers/blob/0cbdd68e14722e911d50eee64ee95823b3ffa09c/src/Utilities/Compiler/Extensions/ISymbolExtensions.cs#L484-L487

```csharp
    private static bool MethodHasCancellationTokenOverload(ITypeSymbol cancellationTokenType, IMethodSymbol method, CancellationToken cancellationToken)
    {
        var overloads = method.ContainingType.GetMembers(method.Name).OfType<IMethodSymbol>();
        return method.GetMatchingOverload(overloads, method.Parameters.Length, cancellationTokenType, cancellationToken) != null;
    }
```

Reply:
Actually you might need to tweak that extension a bit and take an additional parameter for expected parameter count for the overload to find.

Reply:
I didn't do this, hope that's ok.